### PR TITLE
Adds functions to determine if splice rescue can occur for a given variant

### PR DESCRIPTION
**Pull request by ccdlott**

### CODE REVIEW

File: pipeline/splicing/calcVarPriors.py [hunk 1]
```diff
@@ -667,15 +687,292 @@ def getRefAltScores(refSeq, altSeq, donor=False):
...
+            # if variant causes a frameshift, no splice rescue
+            if inFrame == False:
+                priorProb = 0.99
+                spliceRescue = 0
+                frameshift = 1
+            else:
+                varGenPos = int(variant["Pos"])
+                varStrand = getVarStrand(variant)
+                varExonNum = getVarExonNumber(variant)
+                nextExonNum = "exon" + str(int(varExonNum[4:]) + 1)
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: please include comment explaining `nextExonNum = "exon" + str(int(varExonNum[4:]) + 1)`, particularly why index at 4 on

File: pipeline/splicing/calcVarPriors.py [hunk 2]
```diff
@@ -667,15 +687,292 @@ def getRefAltScores(refSeq, altSeq, donor=False):
...
+                refSpliceAccBounds = getRefSpliceAcceptorBoundaries(variant)
+                regionEnd = refSpliceAccBounds[nextExonNum]["acceptorStart"]
+                if varStrand == "-":
+                    regionStart = varGenPos + 8
```
Comment (zfisch) [inline pipeline/splicing/calcVarPriors.py]: include comment for +- 8

File: pipeline/splicing/calcVarPriors.py [hunk 3]
```diff
@@ -382,7 +382,27 @@ def varInExon(variant):
                 if withinBoundaries == True:
                     return True
     return False
-    
+
+def getVarExonNumber(variant):
+    '''
```
[the remark this answers is not in the document]
Comment (ccdlott) [reply pipeline/splicing/calcVarPriors.py]: No this function only works for SNS so far, I changed the name and the function description to match that it only works for SNS variants. I will work on expanding it to variants that span multiple exons later

[the remark this answers is not in the document]
Comment (zfisch) [reply pipeline/splicing/calcVarPriors.py]: 👍

**Comment by ccdlott** [comment]
@melissacline can you please review this to make sure that the logic makes sense from a biological perspective?

**Comment by melissacline** [comment]
I just wanted you to know that I'm working on this.  It will take a while to get through it all, but I should get through it sometime today.

File: pipeline/splicing/calcVarPriors.py [hunk 4]
```diff
@@ -667,15 +686,305 @@ def getRefAltScores(refSeq, altSeq, donor=False):
...
+        varPos -= 1
+        windowStart += 1
+        windowEnd += 1
+        
+    # to get tuple containing sequence with variant position with maximum alt MaxEntScan score
+    maxAltWindowScore = max(windowAltMaxEntScanScores.items(), key=lambda k: k[1])
+    maxVarPosition = maxAltWindowScore[0]
+    maxScores = windowScores[maxVarPosition]
+
+    # determines if variant is in the first three bases of the 9 bp donor sequence
```
Comment (melissacline) [inline pipeline/splicing/calcVarPriors.py]: Again, let's make this a parameter, to minimize the pain when and if we move to a slightly different definition of the critical splice site residues.

File: pipeline/splicing/calcVarPriors.py [hunk 5]
```diff
@@ -667,15 +686,305 @@ def getRefAltScores(refSeq, altSeq, donor=False):
...
+        inFirstThree = True
+            
+    return {"refMaxEntScanScore": maxScores["refMaxEntScanScore"],
+            "refZScore": maxScores["refZScore"],
+            "altMaxEntScanScore": maxScores["altMaxEntScanScore"],
+            "altZScore": maxScores["altZScore"],
+            "varWindowPosition": maxVarPosition,
+            "inFirstThree": inFirstThree}
+
+def varInFirstThree(variant):
```
Comment (melissacline) [inline pipeline/splicing/calcVarPriors.py]: I suggest renaming this to convey the function of this information, to see if the variant is in the exonic portion of the splice site region (which is way too wordy to be a function name verbatim).

File: pipeline/splicing/calcVarPriors.py [hunk 6]
```diff
@@ -667,15 +686,305 @@ def getRefAltScores(refSeq, altSeq, donor=False):
...
+    '''
+    slidingWindowInfo = getMaxMaxEntScanScoreSlidingWindowSNS(variant)
+    if slidingWindowInfo["inFirstThree"] == True:
+        return True
+    return False
+
+def getVarWindowPosition(variant):
+    '''
+    Given a variant, determines window position for highest scoring sliding window
+    Returns integer 1-9 based on variant position with higest scoring window
```
Comment (melissacline) [inline pipeline/splicing/calcVarPriors.py]: Same theme as before.  Rather than "9", indicate the max window size.

File: pipeline/splicing/calcVarPriors.py [hunk 7]
```diff
@@ -667,15 +686,305 @@ def getRefAltScores(refSeq, altSeq, donor=False):
...
+def getNewSplicePosition(varGenPos, varStrand, varWindowPos, inFirstThree):
+    '''
+    Given a variant's:
+    genetic postion, strand, sliding window position with max MES score AND
+      whether that position is within first three bases of highest scoring window
+    Returns the position where splicing occurs for a de novo splice donor
+    '''
+    if varStrand == "+":
+        if inFirstThree == False:
+            newSplicePos = int(varGenPos) - (varWindowPos - 3)
```
Comment (melissacline) [inline pipeline/splicing/calcVarPriors.py]: Same feedback as before.  Rather than using the literal "3", use a parameter to indicate the width of the exonic portion of the splice site window.

File: pipeline/splicing/calcVarPriors.py [hunk 8]
```diff
@@ -667,15 +686,305 @@ def getRefAltScores(refSeq, altSeq, donor=False):
...
+            exonLength = refExonLength - (varExonEnd - (newSplicePos - 1))
+        return exonLength
+
+def compareRefAltExonLengths(refLength, altLength):
+    '''
+    Compares ref and alt exon lengths
+    If both exon lengths % 3 are equal, then both ref and alt have the same reading frame
+    Returns true if both ref and alt exon have the same reading frame, false otherwise
+    '''
+    if refLength % 3 == altLength % 3:
```
Comment (melissacline) [inline pipeline/splicing/calcVarPriors.py]: Love it!  So much complexity reduced to one clean line of code!

File: pipeline/splicing/calcVarPriors.py [hunk 9]
```diff
@@ -667,15 +686,305 @@ def getRefAltScores(refSeq, altSeq, donor=False):
...
+            distanceBetween = wildTypeSplicePos - deNovoSplicePos
+        else:
+            # +1 because of RefSeq numbering
+            distanceBetween = deNovoSplicePos - (wildTypeSplicePos + 1)
+        if distanceBetween % 3 == 0:
+            return True
+        else:
+            return False
+        
+def determineSpliceRescueSNS(variant, boundaries):
```
Comment (melissacline) [inline pipeline/splicing/calcVarPriors.py]: This function name seems off.  As well as determining splice rescue, this function assigns the prior and predicted qualitative class.  How about a broader name?

File: pipeline/splicing/calcVarPriors.py [hunk 10]
```diff
@@ -667,15 +686,305 @@ def getRefAltScores(refSeq, altSeq, donor=False):
...
+                priorProb = 0.99
+                spliceRescue = 0
+                frameshift = 1
+            else:
+                varGenPos = int(variant["Pos"])
+                varStrand = getVarStrand(variant)
+                varExonNum = getVarExonNumberSNS(variant)
+                # varExonNum returns a string in the format "exonN"
+                # nextExonNum parses out N from varExonNum and adds 1 to get next exon number key "exonN+1"
+                nextExonNum = "exon" + str(int(varExonNum[4:]) + 1)
```
Comment (melissacline) [inline pipeline/splicing/calcVarPriors.py]: I'm confused.  What is the significance of the '4'?
